src/build_vehicle_gui.cpp

```diff
@@ -1330,6 +1357,41 @@ struct BuildVehicleWindow : Window {
 		return CargoAndEngineFilter(&item, filter_type);
 	}
 
+	/** Filter by name */
+	bool FilterByName(const Engine *e)
+	{
+		/* Do not filter if the filter text box is empty */
+		if (this->string_filter.IsEmpty()) {
+			return true;
+		}
+
+		char buffer[MAX_LENGTH_VEHICLE_NAME_CHARS * MAX_CHAR_LENGTH];
+		GetString(buffer, e->info.string_id, lastof(buffer));
+		this->string_filter.ResetState();
+		this->string_filter.AddLine(buffer);
+
+		return this->string_filter.GetState();
+	}
+
+	/** Filter by NewGRF extra text */
+	bool FilterByExtraText(const Engine* e)
+	{
+		/* Do not filter if the filter text box is empty */
+		if (this->string_filter.IsEmpty()) {
+			return true;
+		}
+
+		StringID extra_text = GetNewGRFAdditionalText(e->index);
+		if (extra_text != INVALID_STRING_ID) {
+			char buffer[MAX_LENGTH_VEHICLE_NAME_CHARS * MAX_CHAR_LENGTH];
```

Comment:
According to https://github.com/OpenTTD/OpenTTD/pull/10519#issuecomment-1447426417, the max buffer length here appears to be 2046 ASCII characters.

However, the string is likely null-terminated, so the buffer should at least be set to 2048 bytes just in case.